tooling/nargo/src/ops/transform.rs

```diff
@@ -1,21 +1,33 @@
 use acvm::{
     acir::circuit::{ExpressionWidth, Program},
+    compiler::CircuitSimulator,
     FieldElement,
 };
 use iter_extended::vecmap;
-use noirc_driver::{CompiledContract, CompiledProgram};
-use noirc_errors::debug_info::DebugInfo;
+use noirc_driver::{CompilationResult, CompiledContract, CompiledProgram};
+use noirc_errors::{debug_info::DebugInfo, CustomDiagnostic, FileDiagnostic};
 
 pub fn transform_program(
     mut compiled_program: CompiledProgram,
     expression_width: ExpressionWidth,
-) -> CompiledProgram {
+) -> CompilationResult<CompiledProgram> {
     compiled_program.program = transform_program_internal(
         compiled_program.program,
         &mut compiled_program.debug,
         expression_width,
     );
-    compiled_program
+    // Check if the program is solvable
+    for circuit in &mut compiled_program.program.functions {
+        let mut simulator = CircuitSimulator::default();
+        if !simulator.check_circuit(circuit) {
+            let diag = FileDiagnostic {
+                file_id: fm::FileId::dummy(),
+                diagnostic: CustomDiagnostic::from_message("ACVM simulation failed"),
+            };
+            return Err(vec![diag]);
+        }
+    }
+    Ok((compiled_program, Vec::new()))
```

Comment:
On second thoughts, I think we should split this into a separate function which the CLI can run the program through after compilation rather than having it directly integrated into the compilation pipeline.

Reply:
I mean the compilation flow not being compile, transform and then having this step included within transform when you do `nargo compile` it does compile, transform, check.

Thete should be a separate function within the nargo crate and the cli should call that after calling transform.

Reply:
I moved the check into a separate function (and file)